### DISCUSSION TITLE
Flat Config: Create and share multiple configs with different types

### DISCUSSION THREAD

**Discussion by ivodolenc** [eslint/eslint#17414] [Q&A] [answered]
> Patterns specified in `files` and `ignores` use `minimatch` syntax and are evaluated `relative` to the location of the `eslint.config.js` file.

Hi, according to the official [docs](https://eslint.org/docs/latest/use/configure/configuration-files-new#specifying-files-and-ignores), specifying `files` and `ignores` options can be tricky for shared configurations since entries from arrays are relative to a `eslint.config.js` file and not the project `root` which is in most cases desirable.

I believe this is how the `tsconfig.json` file is handled, and according to developer feedback, it's not the best solution.

Let's say we have the configuration from the `npm` package:

```js
// package path: node_modules/package/dist/js/eslint.config.js

export const jsConfig = [
  {
    files: ['src/**/*.js'],
    ignores: ['**/*.tsx']
    // ...
  },
  {
    ignores: ['dir/**/*']
  }
]
```

```js
// project path: ./eslint.config.js

import { jsConfig } from 'npm-package'

export default [...jsConfig]
```

How would that configuration behave? I guess it would be relative to the local project configuration? What if we specify custom config path (`eslint -c ./.config/eslint.config.js .`) that is not in the project root?

Is it best to define `files` and `ignores` separately per package (not a best DX I guess)?

---

Also, if we have a shared configuration with different types (`js` and `ts` for example) without `files` and `ignores`, how can we define it later in a local project?

```js
// package path: node_modules/package/dist/js/eslint.config.js

export const jsConfig = [
  {
    languageOptions: {
      parser: parser,
      parserOptions: {
        //...
      },
      globals: {
        // ...
      }
    },
    rules: {
      // ...
    }
  }
]
```

```js
// package path: node_modules/package/dist/ts/eslint.config.js

export const tsConfig = [
  {
    languageOptions: {
      parser: parser,
      parserOptions: {
        //...
      },
      globals: {
        // ...
      }
    },
    plugins: {
      // ...
    }
    rules: {
      // ...
    }
  }
]
```

Importing that into local project:

```js
// project path: ./eslint.config.js

import { jsConfig, tsConfig } from 'npm-package'

export default [
  ...jsConfig,
  ...tsConfig,
  {
    // how to target jsConfig files and ignores?
    files: ['...'],
    ignores: ['...']
  },
  {
    // how to target tsConfig files and ignores?
    files: ['...'],
    ignores: ['...']
  },
  {
    // global ignores
    ignores: ['...']
  }
]
```

---

Sorry for the long post, but this new system is a bit confusing so I want to better understand how it works and what solutions I have in certain circumstances.

**Comment by aladdin-add** [comment, marked as answer]
> Is it best to define files and ignores separately per package (not a best DX I guess)?

Yes, in general it's a best practice. An example:

```js
import { jsConfig, tsConfig } from 'npm-package'; // assuming exported a config object.


export default [
  {files: ['some-js-files'], ...jsConfig},
  {files: ['some-ts-files'], ...tsConfig},
]
```

**Comment by ivodolenc** [comment]
@aladdin-add Thanks for fast reply.

As Milos suggested in a [comment](https://github.com/eslint/eslint/issues/17400#issuecomment-1647463901), if we want to specify global `ignores` in a shared config, we need to have an array of object.

The first thing that comes to my mind is to store the global `ignores` in a shared config and import them later with the config, but that just complicates things and confuses the users (import the `config` preset and import the global `ignores` separately).

```js
import { jsConfig, tsConfig, globalIgnores } from 'npm-package' // assuming exported a config object.

export default [
  { files: ['some-js-files'], ...jsConfig },
  { files: ['some-ts-files'], ...tsConfig },
  { ignores: globalIgnores } // globalIgnores is an array of strings
]
```

Do you have any advice? Maybe a better workflow/best practice?

**Reply by aladdin-add** [reply]
In the flat config system, the shared config is just a plain js module. This means that how the exports is organized will be the responsibility of the shared config authors, eslint does not force it. The flat config is pretty new, so it may take a while for the community to build what are so-called best practices.

**Reply by aladdin-add** [reply]
In most cases, it is the users' responsibility to ignore some files - share config authors don't know which file should be linted/not linted.

**Reply by ivodolenc** [reply]
I meant something like this:

```js
export const base = {
  ignores: [
    '**/*.d.*',
    '**/node_modules/**/*',
    '**/.git/**/*',
    '**/.private/**/*',
    '**/.cache/**/*',
    '**/dist/**/*'
  ]
}

// and later import into project
import { base } from 'npm-package'
```

Since almost every project has some common dirs that should definitely be ignored (.private, .cache, dist, etc...).

Also, it will recursively ignore all levels so I believe this would work regardless of where the config was located relative to root.

**Reply by aladdin-add** [reply]
sounds great! 👍

**Reply by ivodolenc** [reply]
If we could specify `ignores` inside the object, there would be no complications and confusing situations.

Something like this:

```js
{
  files: ['...'],
  ignoreConfig: ['...'], // this would ignores the current config for the listed entries
  ignores: ['...'] // this would completely ignore the listed entries as any dev would expect :)
},
{
 // ... this config always takes precedence over the previous one
}
```

**Reply by aladdin-add** [reply]
We just need to take some time to let the community get used to it. :)

making changes might cause more confusion.

**Comment by ivodolenc** [comment]
I just tested this and it turns out that we can't import `base`. Can you please look at this @aladdin-add ?

```ts
// npm package

export const base = {
  ignores: ['**/dir/**/*']
}
```

Doesn't work:

```ts
// project eslint.config.js

import { base } from 'npm-package'

export default [
  {
    files: ['...']
    // ...
  },
  {
    ignores: base.ignores
  }
]
```

Works:

```ts
// project eslint.config.js

export default [
  {
    files: ['...']
    // ...
  },
  {
    // when I manually set the same values it works
    ignores: ['**/dir/**/*']
  }
]
```

**Reply by ivodolenc** [reply]
Ah sorry about that, I had to restart the ESLint server in VSCode (vscode-eslint extension). After that it works as expected 👍
